### ISSUE TITLE
Data race in httptest.Server - pkg/storage/etcd, pkg/registry/generic/etcd

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#18928]
The failure is the following:

    WARNING: DATA RACE
    Write by goroutine 95:
      sync.raceWrite()
          /tmp/workdir/go/src/sync/race.go:41 +0x2e
      sync.(*WaitGroup).Wait()
          /tmp/workdir/go/src/sync/waitgroup.go:124 +0xf9
      net/http/httptest.(*Server).Close()
          /tmp/workdir/go/src/net/http/httptest/server.go:168 +0x80
      k8s.io/kubernetes/plugin/pkg/auth/authenticator/token/oidc.TestOIDCAuthentication()
          /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/plugin/pkg/auth/authenticator/token/oidc/oidc_test.go:395 +0x1fc1
      testing.tRunner()
          /tmp/workdir/go/src/testing/testing.go:456 +0xdc

    Previous read by goroutine 142:
      sync.raceRead()
          /tmp/workdir/go/src/sync/race.go:37 +0x2e
      sync.(*WaitGroup).Add()
          /tmp/workdir/go/src/sync/waitgroup.go:66 +0xfa
      net/http/httptest.(*waitGroupHandler).ServeHTTP()
          /tmp/workdir/go/src/net/http/httptest/server.go:198 +0x5c
      net/http.serverHandler.ServeHTTP()
          /tmp/workdir/go/src/net/http/server.go:1862 +0x206
      net/http.(*conn).serve()
          /tmp/workdir/go/src/net/http/server.go:1361 +0x117c

@timothysc @kubernetes/goog-testing @yifan-gu

**Comment by wojtek-t** [comment]
The failure comes from: TestOIDCAuthentication

**Comment by timothysc** [comment]
xref: #18917

**Comment by yifan-gu** [comment]
cc @yifan-gu

**Comment by wojtek-t** [comment]
I've just seen the same error in a different test:

    Write by goroutine 48:
      sync.raceWrite()
          /tmp/workdir/go/src/sync/race.go:41 +0x2e
      sync.(*WaitGroup).Wait()
          /tmp/workdir/go/src/sync/waitgroup.go:124 +0xf9
      net/http/httptest.(*Server).Close()
          /tmp/workdir/go/src/net/http/httptest/server.go:168 +0x80
      k8s.io/kubernetes/pkg/storage/etcd/testing.(*EtcdTestServer).Terminate()
          /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/storage/etcd/testing/utils.go:133 +0x138
      k8s.io/kubernetes/pkg/registry/generic/etcd.TestEtcdWatch()
          /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd_test.go:610 +0xb5f
      testing.tRunner()
          /tmp/workdir/go/src/testing/testing.go:456 +0xdc

    Previous read by goroutine 62:
      sync.raceRead()
          /tmp/workdir/go/src/sync/race.go:37 +0x2e
      sync.(*WaitGroup).Add()
          /tmp/workdir/go/src/sync/waitgroup.go:66 +0xfa
      net/http/httptest.(*waitGroupHandler).ServeHTTP()
          /tmp/workdir/go/src/net/http/httptest/server.go:198 +0x5c
      net/http.serverHandler.ServeHTTP()
          /tmp/workdir/go/src/net/http/server.go:1862 +0x206
      net/http.(*conn).serve()

This failure comes from:
pkg/registry/generic/etcd
TestEtcdWatch

So this may not be related to oidc itself, but maybe to httptest.Server

**Comment by wojtek-t** [comment]
I've also seen exactly the same in pkg/storage/etcd.

**Comment by wojtek-t** [comment]
One more failure is here:
https://00e9e64bacc0329d30a8c72bf3aacbc7eab2390b5b3326030c-apidata.googleusercontent.com/download/storage/v1_internal/b/kubernetes-jenkins/o/logs%2Fkubernetes-test-go%2F4825%2Fartifacts%2Fjunit_v1,extensions-v1beta1_20151220-020118.stdout?qk=AD5uMEu0agHjHt_eE9SdciXlWznADg9yBAO45jdlu8PjKdUKcgnXwXmKYrvnIR8uoVY7BTxf_dw0R5EmJYq0u82zpS-4Q0Fl81fVmj7roGIV5IRwT3aC_FHdh1cY4tn7237SsAa3Bq5OYZtgxDZlvk7V5Bu0P93ffmizdWf1bF-VN6eCa5GDmcnwrbKvXNtCT5yMAWY7sYae_sQwZNuYdqubAHDs1HFWqfxRHjrgtC_7syVAbfwAI_cp9L3xFUOFSr5UIOFiHabtiIl_3d1_b1hgOWlNIoDFgRfHyx7ElI6Z7HRLwNd9LUjC2uQOCL9NY53ImbG_YwRaHA17pIB3NbM6rOVrQpCd7_tv0xKKu8tCeA-F7hDBJF59nSgTjYjHprJkrOZu2d5CbHpUKizJhldtyxDlx6JvdzFitww8ip6Z2Ld2ERTS0QD-1zwtC9fjokmnFaWC1n6tfHQzYViAuI4oYtbBD7RVe3qJwPquapSUWhjBfx553it42hRs2do4GKALGlpTwfwy5O1f4CaWHVqO4OI73RPoeJfWW6zKm7YjgcqVvNIhmkW_3yk8aohOzLdODvieBE0IzkUdxZD3bAg4-SfXo57ejyrJQg_rujhSRq0-6qvkmr2p2aCuPMmJvgOPUO1lOGmUtYTqFumbLoSdztdvUowT9l1MwAqHXdrQSXDmeUNdKkUqmL7Kg0Bt3NQkucxFjiPQwKJbINiKsRb9mItuXdhUdi6wO_X4Ays94mFvkEL49M259nUZIp47Y6uGUL_5g0IRYx6SVu7oSIy3ORER1r_AJ5u0dReDGe7D2NfYZRfQnCgfja0TBDFyI8-UxpxlY09AJ5p_Vjv6BkvlsfNlX9l9OfcrBHygjpkZi0tNkWOHoO6ZXsQK8_xSE48z5YEWSlUDT8Z_80S6UOPO1CZcUck_s9uOrPHmaxQ-1UQ08Z1J10JR952k5Fju11HHftxP7vam

This is an interesting one - it is from TestWatchPurposefulShutdown, and there we already have rt.Gosched, so this will not help.

**Comment by wojtek-t** [comment]
I think I have a hypothesis what can be happening here (although, I'm not able to reproduce it locally).
Basically - it seem for me that there is a race in etcdWatcher (it has always been there, but Go 1.5 exposed it to the world).

What I think is happening here is that:
1. when you call Watch() on etcdHelper, it starts a new go-routine calling etcdWatch()
2. however, there is no synchronization in it - so it can happen that it will be started after some time (e.g. after Stop() is called in the test mentioned above)
3. if that happens, we may be calling etcd server, when it is already terminating...

I will prepare a PR fixing it (although unfortunately I'm not able to reproduce it locally, so this will just be a guess...)

**Comment by wojtek-t** [comment]
I've sent #18956 - although I hope that it will solve most of problems here.

**Comment by wojtek-t** [comment]
However - my PR will not fix OIDC test - I think that one is unrelated...

**Comment by wojtek-t** [comment]
It seems that my fix #18956 didn't help...

**Comment by wojtek-t** [comment]
One more try to fix tests has just been merged #18999 (although I'm still not able to reproduce them locally)

**Comment by wojtek-t** [comment]
Since TestOIDCAuthentication is a different issue - I'm creating a separate issue for that one.

**Comment by wojtek-t** [comment]
Unfortunately the second fix also didn't help (it seems that flakes are rare even after the first fix, but they are still present... - I will debug more.

**Comment by wojtek-t** [comment]
One more attempt is in #19006

**Comment by wojtek-t** [comment]
@fgrzadkowski - hopefully it was finally fixed in the fourth attempt :)
I will keep it open for some more time to monitor it, but it seems that is finally done now.

**Comment by smarterclayton** [comment]
I just saw it https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/2ba3e9748cdcba7f5cb58ef34c697ada2d8c0ba5/kubernetes-pull-test-unit-integration/7258/

**Comment by wojtek-t** [comment]
yeah - I've also seen one flake of pkg/registry/generic/etcd
However, since the last fix I didn't see any pkg/storage/etcd failure.

**Comment by gmarek** [comment]
I've seen one now: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/dec5096bfd0dc6f3c4894a9fe88850efae110825/kubernetes-pull-test-unit-integration/7275/build-log.txt

**Comment by wojtek-t** [comment]
yeah - me too :)

Those flakes are much less frequent now, but they are still there...

**Comment by wojtek-t** [comment]
There are exaclty the same failures coming from pkg/storage, e.g. here:
https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-test-go/5100/artifacts/?project=kubernetes-jenkins

In the above one, the failure is "TestList" from "pkg/storage"

**Comment by wojtek-t** [comment]
[Note that it seems this is also related to watch - in Cacher there is underlying watch to etcd]

**Comment by wojtek-t** [comment]
Ref #19131

I've just run:
$ export KUBE_RACE=-race
$ hack/test-go.sh -i 100 pkg/registry/generic/etcd/

simultanously in 3 consoles - and all 300 tests passed - this is really strange...

**Comment by wojtek-t** [comment]
What is interesting - we are also testing per-resource Watch in registries, e.g. for Pod we have it here:
https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/pod/etcd/etcd_test.go#L288

However - those tests do NOT fail for any single resource.
I have no idea how they are different from tests in pkg/registry/generic/etcd

**Comment by wojtek-t** [comment]
So to summarize the current state:

1. We observe the following races:

        WARNING: DATA RACE
        Write by goroutine 95:
          sync.raceWrite()
              /tmp/workdir/go/src/sync/race.go:41 +0x2e
          sync.(*WaitGroup).Wait()
              /tmp/workdir/go/src/sync/waitgroup.go:124 +0xf9
          net/http/httptest.(*Server).Close()
              /tmp/workdir/go/src/net/http/httptest/server.go:168 +0x80
          k8s.io/kubernetes/plugin/pkg/auth/authenticator/token/oidc.TestOIDCAuthentication()
              /workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/plugin/pkg/auth/authenticator/token/oidc/oidc_test.go:395 +0x1fc1
          testing.tRunner()
              /tmp/workdir/go/src/testing/testing.go:456 +0xdc

        Previous read by goroutine 142:
          sync.raceRead()
              /tmp/workdir/go/src/sync/race.go:37 +0x2e
          sync.(*WaitGroup).Add()
              /tmp/workdir/go/src/sync/waitgroup.go:66 +0xfa
          net/http/httptest.(*waitGroupHandler).ServeHTTP()
              /tmp/workdir/go/src/net/http/httptest/server.go:198 +0x5c
          net/http.serverHandler.ServeHTTP()
              /tmp/workdir/go/src/net/http/server.go:1862 +0x206
          net/http.(*conn).serve()
              /tmp/workdir/go/src/net/http/server.go:1361 +0x117c

2. We observe those in tests in:
  - pkg/registry/generic/etcd
  - pkg/storage
  - pkg/storage/etcd

3. Those races are with Go 1.5.2 (I'm not able to reproduce them locally :()

4. The common factor of all test failures, is that all of them are exercising Watch() - it is __definitely related to Watch__

5. However, not all Watch-related tests are failing - e.g. any of resource-specific Watch-tests from pkg/registry/{resource}/etcd is NOT failing.

If someone has any hints/ideas what can be wrong - please let us know (I'm actually out of any ideas now)

@xiang90 @hongchaodeng

**Comment by xiang90** [comment]
@wojtek-t Can you give me a quick hint about how to reproduce this race? Which unit test should I run? I can try to fix this.

**Comment by wojtek-t** [comment]
@xiang90 - the problem is that I cannot reproduce this locally (I've run those tests 500+ times and all tries passed). On the other hand, when running on Jenkins, it is failing probably at 5-10%.

Currently, the most frequently failing tests isL
pkg/registry/generic/etcd/etcd_test.go
test: TestEtcdWatch

https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/generic/etcd/etcd_test.go#L560

To be honest, I suspect the issue is somewhere in Kubernetes code (pkg/storage/etcd/etcd_watcher.go ?), but I don't have any proof and/or suspicions.

**Comment by xiang90** [comment]
@wojtek-t Oh. No worries. etcd or k8s do not really matter.

How can I reproduce on Jenkins? I think I need to reproduce this so that I can find the root cause more easily and know I actually fix it.

**Comment by wojtek-t** [comment]
@xiang90 - you can try writing "@k8s-bot unit test this please" few times - if you run 10 times successfully (without those those unit test failure), it would probably mean that the test is fixed.
Once you have WIP PR, I'm happy to help with testing it.

**Comment by yifan-gu** [comment]
@wojtek-t The oidc test is still failing? I thought it was fixed by https://github.com/kubernetes/kubernetes/pull/18917

**Comment by wojtek-t** [comment]
@yifan-gu - no - it seems to be fixed (I've already changed the title of this issue)

**Comment by yifan-gu** [comment]
@wojtek-t That's cool :) Sorry, I saw your [comments](https://github.com/kubernetes/kubernetes/issues/18928#issuecomment-167574597) on summarizing the current state 6 indicate the oidc test is still failing recently, so I got confused. Never mind.

**Comment by wojtek-t** [comment]
@yifan-gu - sorry; I pasted the wrong trace. it's already updated

**Comment by wojtek-t** [comment]
@xiang90 - did you do any progress on it?

**Comment by xiang90** [comment]
@wojtek-t I will do it tomorrow. Sorry for the delay.

**Comment by wojtek-t** [comment]
@xiang90 - no worries; I will also try to do something today - will let you know if I find something

**Comment by wojtek-t** [comment]
I added some logging and it seems that we still have some unwanted requests.
The PR adding logging is here: #19158
the failure log is here:
https://gist.github.com/wojtek-t/9f5e23b714b876c140f7

I will continue debugging.

**Comment by wojtek-t** [comment]
The line that is suspicious is the "Get" request after the "data race"

**Comment by wojtek-t** [comment]
One more failure coming from TestList in pkg/storage/cacher_test.go
https://gist.github.com/wojtek-t/2328d11f9d7bc72642a0

To be honest, I have no idea where does this race come from...
@xiang90 - it would be great if you could look into it (this issue is causing us a lot of problems)

**Comment by xiang90** [comment]
@wojtek-t The race comes from go httptest lib itself. See https://github.com/golang/go/issues/12262. It is still in go 1.5 but fixed in go tip.

I do not know what should we do actually. Do you want us to work around this?

**Comment by wojtek-t** [comment]
@xiang90 - thanks a lot for investigating it.
It would be great it if we could somehow work around it, because it is often blocking our merge queue.

Since this is just test code, maybe some "sleep" before closing etcd server would be enough?

**Comment by xiang90** [comment]
@wojtek-t It is not about etcd server. There is a httptest server. Accepting requests while closing will lead to the race. We need to ensure there is not more NEW request coming before close the httptest server. So probably we should stop all clients, wait for a while (sleeping...) and then close the server.

**Comment by wojtek-t** [comment]
Yes - that's what I had in mind.
All the tests are showing that the issue is only watch requests.
What I've done here:
https://github.com/kubernetes/kubernetes/blob/master/pkg/storage/etcd/etcd_watcher.go#L472
was supposed to ensure that no new requests will be issued to httptest server.

Also, all watchers is tests should now be explicitly Stopped() before the end of the test (before closing of httptest server and etcd is done).

So what I was asking was whether adding some Sleep right before closing httptest server (when all clients are supposed to be already closed) should help?

**Comment by xiang90** [comment]
@wojtek-t 

I checked the log here: https://gist.github.com/wojtek-t/9f5e23b714b876c140f7

I expect to see https://github.com/kubernetes/kubernetes/pull/19158/files#diff-3873aaf6d0eff7554d081a4b7bebeae3R463 printed out before `closing client connections`. Seem like we fail to close the watcher before close server?

**Comment by xiang90** [comment]
@wojtek-t Also checked the stop code in https://github.com/kubernetes/kubernetes/blob/master/pkg/storage/etcd/etcd_watcher.go. I feel it is OK.

**Comment by wojtek-t** [comment]
The first link that I pasted was before I added this "closing watcher" log - I realized that it would be needed later and after that the only failure that I reproduced was the second one (where the logs seem ok)

**Comment by xiang90** [comment]
@wojtek-t So the only issue is the TestList one right now?

**Comment by xiang90** [comment]
@wojtek-t Basically what happened is that we cancel ctx passed into etcd.watch. But in etcd client the cancellation is done asyn (standard library only provides async http request cancellation). So you can see a get request after next is finished. A quick fix would be just sleep for a while.

**Comment by xiang90** [comment]
If you want to be sure, probably you need to provide a customized transporter to etcd which ensures the http cancellation is sync.

**Comment by wojtek-t** [comment]
@xiang90 

>
@wojtek-t So the only issue is the TestList one right now?

No - there are other tests failing too, but all of them are failing with exactly the same trace.

>
@wojtek-t Basically what happened is that we cancel ctx passed into etcd.watch. But in etcd client the cancellation is done asyn (standard library only provides async http request cancellation). So you can see a get request after next is finished. A quick fix would be just sleep for a while.

hmm - I'm not sure I understand.
Because in etcdWatcher::Stop(), I'm waiting until the watch is really finished using WatiGroup:
https://github.com/kubernetes/kubernetes/blob/master/pkg/storage/etcd/etcd_watcher.go#L472

And in watch loop, that WaitGroup would be fullfilled, only when the Next() call is already finished:
https://github.com/kubernetes/kubernetes/blob/master/pkg/storage/etcd/etcd_watcher.go#L189

So what you are saying is that, even though the Next() call is already finished, etcd client library can still issue some request to etcd?

**Comment by wojtek-t** [comment]
It seems that answer for my question above:

>
So what you are saying is that, even though the Next() call is already finished, etcd client library can still issue some request to etcd?

is yes.

For me this seems like a real bug (https://github.com/golang/go/issues/12262 is unrelated to that).

I think that we have 3 options:
1. add Sleep() just before terminating http server:
https://github.com/kubernetes/kubernetes/blob/master/pkg/storage/etcd/testing/utils.go#L130
2. disable race checking in those tests
3. (what @xiang90 suggested): "provide a customized transporter to etcd client configuration which ensures the http cancellation is sync"

I definitely think that 3 is what we should do. @xiang90 would you be able to do it (or find someone who would do it)? [I'm leaving for paternity leave, so won't be able to do it]

**Comment by wojtek-t** [comment]
@xiang90 - or even better, would it be possible to guarantee that when Next() finishes, no http request will be issues to etcd?

**Comment by xiang90** [comment]
@wojtek-t 

> @xiang90 - or even better, would it be possible to guarantee that when Next() finishes, no http request will be issues to etcd?

That requires etcd/client to count the number of in-flight HTTP requests. Not sure if that worth the effort in general. We have no way to know if the http request is sent out. The HTTP cancellation in golang is just a hint by default and will return immediately. 

I am not sure why you even want to care about that other than the httptest server will report a race due to its internal implementation issue.

**Comment by wojtek-t** [comment]
What I would like to avoid is the situation when Next() finishes and still we can issue some request.

**Comment by wojtek-t** [comment]
But yeah - if can solve it with passing a different transport, I'm fine with that.

**Comment by xiang90** [comment]
> What I would like to avoid is the situation when Next() finishes and still we can issue some request.

Well... We want to avoid the other side over the network receiving a request after cancel of request returns. This is not what normal transport implementation wants to achieve. It means we have to wait the ack from the other side once the request sent out over the network. For the cancellation, normally we just want to unblock the sender from sending or waiting.

**Comment by wojtek-t** [comment]
Sure, I agree. But we should be able to gracefully terminate the test - currently we can't do it.

**Comment by xiang90** [comment]
@wojtek-t For that purpose, I would suggest to have a transport(https://golang.org/pkg/net/http/#Transport) that counts the current opening connection by overwrite the dial func, add a tr.Terminate funcs that close all ongoing connections and also stop accepting new ones. That should not be hard to do.

**Comment by wojtek-t** [comment]
@xiang90 - would you be able to do that? I'm just leaving for paternity leave and will be OOO for ~month.

**Comment by xiang90** [comment]
@wojtek-t OK. I will try.

**Comment by wojtek-t** [comment]
@xiang90 Thanks a lot!
@gmarek please help @xiang90 with reviewing, etc.

**Comment by yifan-gu** [comment]
@wojtek-t Congrats!

**Comment by wojtek-t** [comment]
@yifan-gu thanks!

**Comment by quinton-hoole** [comment]
Saw this again in https://github.com/kubernetes/kubernetes/pull/17733#issuecomment-168755295

**Comment by gmarek** [comment]
I'll merge #19187 to make this flakes go away.

**Comment by gmarek** [comment]
If anyone see this happening again please ping this issue.

**Comment by davidopp** [comment]
@gmarek should we close this since you merged #19187? if so, please close.

**Comment by davidopp** [comment]
Oh sorry, I see @xiang90 is working on a less hacky solution, as described here
https://github.com/kubernetes/kubernetes/issues/18928#issuecomment-168055281

Re-assigning to him.

**Comment by davidopp** [comment]
But if we think @wojtek-t 's sleep will fix the problem good enough for now, we should lower the priority of this from P0 to P1 or P2...

**Comment by xiang90** [comment]
@davidopp Right. Agree to de-prioritize it. I will try to get a more robust one next week.

**Comment by davidopp** [comment]
Thanks, bumped to P2.

**Comment by ihmccreery** [comment]
@xiang90 We're prioritizing flakes for the time being.  Is this something you'll be able to get to soon, or should we find someone else to work on it?

**Comment by xiang90** [comment]
@ihmccreery I am in China recently. I can look into this later. But I assume this issue has been fixed by a work around. And go will also fix the root cause soon. This issue is just to make the test shutdown in a more gracefully way.

**Comment by ihmccreery** [comment]
Data race in `net/http/test` in GoLang that will not fix.

**Comment by davidopp** [comment]
As @brendandburns mentioned, this might be fixed by #19458

**Comment by ihmccreery** [comment]
Reassigning to @davidopp based on #19458.

**Comment by timothysc** [comment]
iirc the long term fix is in etcd proper.

**Comment by brendandburns** [comment]
I'm taking this due to #20123 having the fix.

**Comment by xiang90** [comment]
@timothysc etcd/client does not really manage the transport layer. We generate and pass in the transport in k8s and k8s has full management of the shutdown. We need to change that part.

**Comment by brendandburns** [comment]
I'm going to close this issue, since I beleive that #20123 fixes the original issue in the title.

@wojtek-t @xiang90 if you want to open a separate issue related to what you are seeing in etcd, please file a separate issue.
